---
title: "Workflow of Genome Range Assignment_Jin (Ching-Yi Liao)"
output: html_document
---

Sildes 132-139 of the Genomic Ranges lecture
Buffalo book Chapter 9, pp 325-327

##Set up##

```{r}
source("http://bioconductor.org/biocLite.R")
biocLite()
library(GenomicRanges)
library(BiocInstaller)
biocLite("GenomicFeatures")
biocLite("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(rtracklayer)

txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene
txdb
```

# Extract variants by exons that are represented by genomic ranges#

#1. Import a file with variants (SNPs, indels, etc...) from chr1 of *Mus musculus*#

```{r}
setwd("C:/Users/user/bds-files/chapter-09-working-with-range-data/")
dbsnp137 <- import("mm10_snp137_chr1_trunc.bed.gz")
```

#2. Collapse all overlapping exons in the mouse TranscriptDb object#

```{r}
collapsed_exons <- reduce(exons(txdb), ignore.strand=TRUE) #extract and collapse all overlapping exons with reduce()
chr1_collapsed_exons <- collapsed_exons[seqnames(collapsed_exons) == "chr1"]
```

#3. Inspect the variant file before extracting varients in exons#

```{r}
summary(width(dbsnp137))
```

#4. Adjust the width of variants#

```{r}
dbsnp137_resized <- dbsnp137
zw_i <- width(dbsnp137_resized) == 0
dbsnp137_resized[zw_i] <- resize(dbsnp137_resized[zw_i], width=1)
```

#5. Pull out variants that overlap exons on chromosome 1#

```{r}
hits <- findOverlaps(dbsnp137_resized, chr1_collapsed_exons, 
        ignore.strand=TRUE)
```

#6. Determine the number of variants and the proportion of variants that are exonic#

```{r}
length(unique(queryHits(hits))) 
length(unique(queryHits(hits)))/length(dbsnp137_resized)
```

#7. Find the number of variants per exon#

```{r}
var_counts <- countOverlaps(chr1_collapsed_exons, dbsnp137_resized, ignore.strand=TRUE)
```

#8. Append to the GRanges object that includes exons#

```{r}
chr1_collapsed_exons$num_vars <- var_counts 
```

#9. File output#

```{r}
setwd("C:/Users/user/Ranges-Assignment/")
write.table(chr1_collapsed_exons, file = "NumberVariantsperExon.txt", row.names = FALSE) #write a .txt file 
```

## Bonus##

Buffalo book Chapter 9, pp 314-319

#Retrieving promoter regions#

#1. Set up#

```{r}
setwd("C:/Users/user/bds-files/chapter-09-working-with-range-data/")
mm_gtf <- import('Mus_musculus.GRCm38.75_chr1.gtf.gz')
```

#2. Flank and promoters#
(1) Select protein coding genes by subsetting the **gene_biotype** column.
```{r}
table(mm_gtf$gene_biotype)
```
(2) Subset all features that have type "gene" and biotype "protein_coding."
```{r}
chr1_pcg <- mm_gtf[mm_gtf$type == "gene" &
mm_gtf$gene_biotype == "protein_coding"]

summary(width(chr1_pcg))

length(chr1_pcg) #check the length distribution 
```

(3) Grab 3kbp upstream of each feature.
```{r}
chr1_pcg_3kb_up <- flank(chr1_pcg, width=3000)

chr1_pcg_3kb_up2 <- promoters(chr1_pcg, upstream=3000, downstream=0) #mimic flank() call
identical(chr1_pcg_3kb_up, chr1_pcg_3kb_up2) #show the result
```

#3. Connection genomic ranges with sequence data#

```{r}
library(BiocInstaller)
biocLite("BSgenome")
biocLite("BSgenome.Mmusculus.UCSC.mm10")

library(BSgenome.Mmusculus.UCSC.mm10)
mm_gm <- BSgenome.Mmusculus.UCSC.mm10
organism(mm_gm) #accessor function to extract information from BSgenome packages
providerVersion(mm_gm) #accessor function to extract information from BSgenome packages
provider(mm_gm) #accessor function to extract information from BSgenome packages

seqinfo(mm_gm) #look at sequence information
mm_gm$chrM
mm_gm[[22]]

library(Biostrings)
matchPattern("GGCGCGCC", mm_gm$chr1) #BSgenome objects can be searched using the string-matching and alignment functions
```

(1) Check all sequences that will be grabbed are in the BSgenome object.
```{r}
all(seqlevels(chr1_pcg_3kb_up) %in% seqlevels(mm_gm))
```
(2) Create a test **GRanges** object to show how we can manually change chromosome names.
```{r}
gr <- GRanges(c("chr1", "chr2"), IRanges(start=c(3, 4), width=10))
```
(3) Change the sequence levels using **seqlevels()**.
```{r}
seqlevels(gr) <- c("1", "2")
seqlevels(gr)
```
(4) Use **seqlevelsStyle()** to switch between the style "chr1" (UCSC style) and "1" (Ensembl/NCBI style)
```{r}
seqlevelsStyle(chr1_pcg_3kb_up)
seqlevelsStyle(mm_gm)
```
(5) Set the style to "UCSC"
```{r}
seqlevelsStyle(chr1_pcg_3kb_up) <- "UCSC"
all(seqlevels(chr1_pcg_3kb_up) %in% seqlevels(mm_gm))
```

Grab the sequences for particular regions kept in a **GRanges** object.
```{r}
chr1_3kb_seqs <- getSeq(mm_gm, chr1_pcg_3kb_up)
chr1_3kb_seqs
```

#4. File output#

Write the sequences to a FASTA file.
```{r}
setwd("C:/Users/user/Ranges-Assignment/")
writeXStringSet(chr1_3kb_seqs, file="mm10_chr1_3kb_promoters.fasta",
format="fasta")
```
